flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/FunctionDefinitionFactory.java

```diff
@@ -34,4 +34,29 @@ public interface FunctionDefinitionFactory {
      * @return a {@link FunctionDefinition}
      */
     FunctionDefinition createFunctionDefinition(String name, CatalogFunction catalogFunction);
+
+    /**
+     * Creates a {@link FunctionDefinition} from given {@link CatalogFunction} with the given {@link
+     * Context} containing the class loader of the current session, which is useful when it's needed
+     * to load class from class name.
+     *
+     * <p>The default implementation will call {@link #createFunctionDefinition(String,
+     * CatalogFunction)} directly.
+     *
+     * @param name name of the {@link CatalogFunction}
+     * @param catalogFunction the catalog function
+     * @param context the {@link Context} for creating function definition
+     * @return a {@link FunctionDefinition}
+     */
+    default FunctionDefinition createFunctionDefinition(
+            String name, CatalogFunction catalogFunction, Context context) {
+        return createFunctionDefinition(name, catalogFunction);
```

Comment:
When calling the old method, we should replace the thread classloader with the user classloader (see `org.apache.flink.table.client.gateway.context.ExecutionContext#wrapClassLoader`). Otherwise, classes can't be found if it still use the old method. Please add a test in the planner to cover this problem.

Reply:
Add a test in [CatalogTableITCase.scala](https://github.com/apache/flink/pull/20211/files#diff-ec8b5b6a6d4b54e45f17bfe66e232ee256553d5e3d4aff16d34b72b6c4373ecb)